src/anonymous-credentials/schema.ts

```diff
@@ -587,56 +624,78 @@ export class CredentialSchema extends Versioned {
+   * Currently, does not check if the needed JSON-schema definitions are actually present but assumes that they will be
+   * already passed.
+   * @param node
+   * @param nodeKeyName - Name of the node, used for throwing more informative error message
+   */
+  static convertToInternalSchemaObj(node: any, nodeKeyName: string = ''): object {
+    // util function needed only in this func
+    const createFullName = (old: string, neww: string): string => {
+      return old.length == 0 ? neww : `${old}.${neww}`;
+    };
 
-    if (node.type === 'object') {
-      if (typeof node.properties !== 'undefined') {
-        const result: object = {};
-        const keys = Object.keys(node.properties);
-        keys.forEach((k) => {
-          const value = node.properties[k];
-
-          let insVal = {};
-          if (value.type === 'object') {
-            if (typeof value.properties === 'object') {
-              insVal = CredentialSchema.flattenJSONSchema(value);
-            } else {
-              throw new Error(`${k} must have properties field that is an object`);
-            }
-          } else if (value.type === 'array') {
-            if (Array.isArray(value.items)) {
-              insVal = value.items.map((i) => CredentialSchema.flattenJSONSchema(i));
-            } else {
-              throw new Error('No indefinite length array support');
-            }
+    // Will either have a "type" property or will be defined using "$ref"
+    const typ = node.type;
+    const ref = node.$ref;
+    if (typ !== undefined) {
+      switch (typ) {
+        case 'string':
+          return node;
+        case 'integer':
+          // If key "minimum" is not present or not an integer, validation logic in other function will throw.
+          return { type: this.INT_TYPE, minimum: node.minimum };
+        case 'number':
+          return { type: this.NUM_TYPE, minimum: node.minimum, decimalPlaces: getDecimalPlaces(node.multipleOf) };
+        case 'object':
+          if (node.properties !== undefined) {
+            const result = {};
+            Object.entries(node.properties).forEach(([k, v]) => {
+              result[k] = CredentialSchema.convertToInternalSchemaObj(v, createFullName(nodeKeyName, k));
+            });
+            return result;
           } else {
-            insVal = value;
+            throw new Error(`Schema object key ${nodeKeyName} must have properties object`);
           }
-
-          result[k] = insVal;
-        });
-        return result;
-      } else {
-        throw new Error('Schema object must have properties object');
+        case 'array':
+          return node.items.map((i) => CredentialSchema.convertToInternalSchemaObj(i, createFullName(nodeKeyName, i)));
+        default:
+          throw new Error(`Unknown type for key ${nodeKeyName} in schema: ${typ}`);
+      }
+    } else if (ref !== undefined) {
+      switch (ref) {
```

Comment:
This is a bad way to handle refs, refs should be defined int the schema and expanded upon parsing

Reply:
> but you should not hard code them

There is no in-built type for encryptable/reversible so I don't know how else would you represent and parse it without defining a convention. Do you have an alternate in mind? 
Secondly, this was the same thing I had described when i shared the example schema in slack.

Reply:
>There is no in-built type for encryptable/reversible so I don't know how else would you represent and parse it without defining a convention. Do you have an alternate in mind?

no the definition part is fine, i have no problem with how you approached that. my concern was in ref handling (the next few lines in this method that was highlighted). please see commit below

> this was the same thing I had described when i shared the example schema in slack

yeah and to state again - no issues with that approach at all. think you misunderstood my comment. let me elaborate...

my problem related to what i mentioned with `allOf` - allowing them to use refs but only specific refs could be confusing. it makes people (potentially) believe that they can use refs in their schema - when in reality they can only use 3 hard defined refs. what i envisioned is that you would support refs using json pointers and overriding definitions (partly you did this). it wasnt too much extra effort to support refs in their entirity here.

check this commit: https://github.com/docknetwork/crypto-wasm-ts/pull/5/commits/659dd55dd74b7a18d7fcb516a706ee845eff0bec in this branch for how i envisioned it being handled. its similar to your way, but now supports expanding refs defined by the user and is a cleaner way of "hard coding" them rather than the non-modifiable switch check. now thats its a static attribute, if needed someone can modify that and also read it to understand the overriding definitions

also added test for user defined refs as they can be quite useful when defining schemas in general